backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/attribute/impl/HUUniqueAttributesRepository.java

```diff
@@ -0,0 +1,136 @@
+
+	public void createOrUpdateHUUniqueAttribute(@NonNull final HUUniqueAttributeCreateRequest huUniqueAttributeCreateRequest)
+	{
+		final I_M_HU_UniqueAttribute existingHUUniqueAttribute = retrieveHUUniqueAttributeForHU(huUniqueAttributeCreateRequest.getHuId());
+
+		if (existingHUUniqueAttribute != null)
+		{
+			// hu was unique attribute was already recorded. Update the value if needed
+			if (!huUniqueAttributeCreateRequest.getAttributeValue().equals(existingHUUniqueAttribute.getValue()))
+			{
+				existingHUUniqueAttribute.setValue(huUniqueAttributeCreateRequest.getAttributeValue());
+				save(existingHUUniqueAttribute);
+			}
+
+			return;
+		}
+
+		final I_M_HU_UniqueAttribute huUniqueAttributeRecord = newInstance(I_M_HU_UniqueAttribute.class);
+
+		huUniqueAttributeRecord.setM_HU_PI_Attribute_ID(huUniqueAttributeCreateRequest.getHuPIAttributeId().getRepoId());
+		huUniqueAttributeRecord.setM_HU_Attribute_ID(huUniqueAttributeCreateRequest.getHuAttributeId());
+		huUniqueAttributeRecord.setM_HU_ID(huUniqueAttributeCreateRequest.getHuId().getRepoId());
+		huUniqueAttributeRecord.setM_Product_ID(huUniqueAttributeCreateRequest.getProductId().getRepoId());
+		huUniqueAttributeRecord.setM_Attribute_ID(huUniqueAttributeCreateRequest.getAttributeId().getRepoId());
+		huUniqueAttributeRecord.setValue(huUniqueAttributeCreateRequest.getAttributeValue());
+
+		save(huUniqueAttributeRecord);
+	}
+
+	private I_M_HU_UniqueAttribute retrieveHUUniqueAttributeForHU(@NonNull final HuId huId)
+	{
+		return queryBL.createQueryBuilder(I_M_HU_UniqueAttribute.class)
+				.addEqualsFilter(I_M_HU_UniqueAttribute.COLUMNNAME_M_HU_ID, huId)
+				.create()
+				.firstOnly(I_M_HU_UniqueAttribute.class);
+	}
+
+	public I_M_HU_UniqueAttribute retrieveHUUniqueAttributeForProductAndAttributeValue(@NonNull final HUUniqueAttributeParameters huUniqueAttributeParameters)
+	{
+		final IQueryBuilder<I_M_HU_UniqueAttribute> queryBuilder = queryBL.createQueryBuilder(I_M_HU_UniqueAttribute.class)
+				.addEqualsFilter(I_M_HU_UniqueAttribute.COLUMNNAME_M_Product_ID, huUniqueAttributeParameters.getProductId())
+				.addEqualsFilter(I_M_HU_UniqueAttribute.COLUMNNAME_M_Attribute_ID, huUniqueAttributeParameters.getAttributeId())
+				.addEqualsFilter(I_M_HU_UniqueAttribute.COLUMNNAME_Value, huUniqueAttributeParameters.getAttributeValue());
+
+		if (huUniqueAttributeParameters.getHuIdToIgnore() != null)
+		{
+			queryBuilder.addNotEqualsFilter(I_M_HU_UniqueAttribute.COLUMNNAME_M_HU_ID, huUniqueAttributeParameters.getHuIdToIgnore());
+		}
+		return queryBuilder
+				.create()
+				.firstOnly(I_M_HU_UniqueAttribute.class);
+	}
+
+	public List<I_M_HU_Attribute> retrieveHUAttributes(@NonNull final HuPackingInstructionsAttributeId huPIAttributeId)
+	{
+		return queryBL.createQueryBuilder(I_M_HU_Attribute.class)
+				.addEqualsFilter(I_M_HU_Attribute.COLUMNNAME_M_HU_PI_Attribute_ID, huPIAttributeId)
+				.create()
+				.list(I_M_HU_Attribute.class);
```

Comment:
there might be a 110million M_HU_Attributes. Even if this is rarely changed, please return an Iterator to avoid OOMEs (i think having the user wait for quite some time is justifyable - they will du this very rearly and will know what they do)